src/promptflow-devkit/promptflow/batch/_batch_engine.py

```diff
@@ -509,11 +511,21 @@ async def _exec_batch(
         batch_inputs: List[Mapping[str, Any]],
         run_id: Optional[str] = None,
     ) -> List[LineResult]:
+        # line_results as input parameter, so that the completed line results can be summarized
+        # when batch run is canceled.
         worker_count = self._worker_count or DEFAULT_CONCURRENCY
         semaphore = asyncio.Semaphore(worker_count)
+
+        batch_inputs_dict = {}
+        if any(LINE_NUMBER_KEY not in input for input in batch_inputs):
+            bulk_logger.warning("There are lines without line number in the batch inputs, use index as line number.")
+            batch_inputs_dict = {i: input for i, input in enumerate(batch_inputs)}
+        else:
+            batch_inputs_dict = {input[LINE_NUMBER_KEY]: input for input in batch_inputs}
+
```

Comment:
Doesn’t local python and single container need this logic? Should it all be consolidated in the code shown in the figure below?
![image](https://github.com/microsoft/promptflow/assets/111329184/0cbf0b89-6690-40cd-833b-79734b5f3e19)

Reply:
Offline synced. There should be no case where there's no LINE_NUBMER_KEY in batch input.
remove the first if.